command/agent/consul/namespaces_client.go

```diff
@@ -0,0 +1,42 @@
+package consul
+
+import (
+	"sort"
+	"strings"
+)
+
+// NamespacesClient is a wrapper for the Consul NamespacesAPI, that is used to
+// deal with Consul OSS vs Consul Enterprise behavior in listing namespaces.
+type NamespacesClient struct {
+	namespacesAPI NamespaceAPI
+}
+
+// NewNamespacesClient returns a NamespacesClient backed by a NamespaceAPI.
+func NewNamespacesClient(namespacesAPI NamespaceAPI) *NamespacesClient {
+	return &NamespacesClient{
+		namespacesAPI: namespacesAPI,
+	}
+}
+
+// List returns a list of Consul Namespaces.
+//
+// If using Consul OSS, the list is a single element with the "default" namespace,
+// even though the response from Consul OSS is an error.
+func (ns *NamespacesClient) List() ([]string, error) {
+	namespaces, _, err := ns.namespacesAPI.List(nil)
+	if err != nil {
+		// check if the error was a 404, indicating Consul is the OSS version
+		// which does not have the /v1/namespace handler
+		if strings.Contains(err.Error(), "response code: 404") {
```

Comment:
This should change before 1.0 GA; instead we can fingerprint Consul oss/ent and set a boolean to do the Right Thing without leaning on Consul error response.